pallets/loan/src/lib.rs

```diff
@@ -300,37 +331,72 @@ pub mod pallet {
 			Ok(())
 		}
 
-		/// Closes a given loan if repaid fully
-		#[pallet::weight(100_000)]
+		/// Closes a given loan
+		///
+		/// Loan can be closed on two scenarios
+		/// 1. When the outstanding is fully paid off
+		/// 2. When loan is written off 100%
+		/// Loan status is moved to Closed
+		/// Asset/Collateral is transferred back to the loan owner.
+		/// LoanNFT is transferred back to LoanAccount.
+		#[pallet::weight(
+			<T as Config>::WeightInfo::repay_and_close().max(
+				<T as Config>::WeightInfo::write_off_and_close()
+			)
+		)]
 		#[transactional]
 		pub fn close_loan(
 			origin: OriginFor<T>,
 			pool_id: PoolIdOf<T>,
 			loan_id: T::LoanId,
-		) -> DispatchResult {
+		) -> DispatchResultWithPostInfo {
 			let owner = ensure_signed(origin)?;
-			let asset = Self::close(pool_id, loan_id, owner)?;
+			let (asset, written_off) = Self::close(pool_id, loan_id, owner)?;
 			Self::deposit_event(Event::<T>::LoanClosed(pool_id, loan_id, asset));
-			Ok(())
+			match written_off {
+				true => Ok(Some(T::WeightInfo::write_off_and_close()).into()),
+				false => Ok(Some(T::WeightInfo::repay_and_close()).into()),
+			}
```

Comment:
`nit`

To avoid the noise of mixing the logic of the different behaviours taken for when `written_off` is true or false, and creating the result value in the same line, we could separate the steps:

``` rust
      let collateral = match written_off {
				true => T::WeightInfo::write_off_and_close(),
				false => T::WeightInfo::repay_and_close(),
			};

			Ok(Some(collateral).into())
```

Reply:
not sure I follow your suggestion implementation

Reply:
My idea is to separate the step 1) where we calculate the output asset/collateral based on the value of `written_off` from 2) the step of building and returning a valid output value for the function itself. I left a `nit` flag so feel free to ignore this, it's just that _for me_ mixing the two things in one step makes is harder to distinguish these two separate things.

Reply:
🤔